Add env and file options to extend support for different injection mechanisms

mr-beerkiss opened the pull request:
- ~added support for `env:` and `file:` prefixes on the `token` field to support different mechanisms of reading the token value.~
- Added support for `env` and `file` parameters to support different mechanisms of supplying the token

mr-beerkiss committed:
Support env: and file: prefixes

mr-beerkiss committed:
Update README.md

mattcompiles commented on README.md:
```diff
@@ -3,39 +3,77 @@
-      seek-oss/private-npm#v1.0.1:
-        token: ${NPM_TOKEN}
+      seek-oss/private-npm#v1.1.1:
+        token: ${MY_TOKEN}
```
Maybe update this to the preferred approach?

mr-beerkiss replied:
Not sure I follow you here, I think this is the variant we wish to deprecate in favour of `env` or `file`

mattcompiles replied:
This is the section in the docs talking about custom npm registries. We should reference the non deprecated version of the API here.

mr-beerkiss replied:
Oh right I'm with you now.  Good point I'll fix that.

mattcompiles commented on README.md:
```diff
@@ -3,39 +3,77 @@
 
 ### `token` (required)
-The value of the NPM token.  
+The value of the NPM token.  Using the prefix `env:` of `file:` changes the behaviour of how the token is read.  By 
```
Using the prefix `env:` or `file:`

mattcompiles reviewed (commented):
Looks good mate. Just a couple of typos.

zsims commented on README.md:
```diff
@@ -3,39 +3,77 @@
-      seek-oss/private-npm#v1.0.1:
-        token: ${NPM_TOKEN}
+      seek-oss/private-npm#v1.1.1:
+        token: "env:MY_TOKEN"
```
How do you disambiguate tokens vs parameters for the token?

If you do this via YAML then you can have the best of all worlds:

`BUILDKITE_PLUGIN_PRIVATE_NPM_TOKEN`:
```
  seek-oss/private-npm#v1.1.1:
    token: "hard coded"
```

`BUILDKITE_PLUGIN_PRIVATE_NPM_TOKEN_ENV`:
```
  seek-oss/private-npm#v1.1.1:
    token:
      env: MY_TOKEN
```

`BUILDKITE_PLUGIN_PRIVATE_NPM_TOKEN_FILE`:
```
  seek-oss/private-npm#v1.1.1:
    token:
      file: "my-file"
```

mr-beerkiss replied:
Yep that was my first instinct, but having three approaches means I can't make any one of them required in the `plugin.yml` file.  I think this is a better approach so perhaps it worth the trade off of having that extra layer of validation enforced by Buildkite

zsims replied:
I still think it's worth doing. None of the plugins seem to do "deep" schemas (e.g. what's in an `array` or that `string` can't be empty). The approach in this PR relies on inspecting the opaque token.

You can also do validation at plugin execution time, e.g. that one of those three values is required

zsims replied:
IMO it's worth doing:

```
  properties:
    registry:
      type: string
    token:
      type: [string, object]
      properties:
        env:
          type: string
        file:
          type: string
``` 

and some basic validation in bash (e.g. one required). I really don't think the approach here is consistent with other Buildkite plugins.

mr-beerkiss committed:
Negate prefixes in favour of distintive fields

mr-beerkiss committed:
Fix default var assignments

mr-beerkiss committed:
Force exclusive use of fields

mr-beerkiss committed:
Updated README

mr-beerkiss commented:
@zsims & @mattsjones I've implemented the changes suggested by @zsims to have explicit fields for `env`, `file` and `token` behaviours.  Updated tests and docs accordingly

zsims commented on README.md:
```diff
@@ -3,39 +3,84 @@
+Example: `${MY_TOKEN}`
+> **NOTE:** Don't put your tokens into source control.  Don't use web interfaces you don't control to inject them into 
+> your environment either.  Rather use a Secrets Manager.  If you are an AWS user, perhaps consider the 
+> [aws-sm-buildkite-plugin](https://github.com/seek-oss/aws-sm-buildkite-plugin) which works well with this plugin.
```
Nice!

zsims reviewed (approved):
Nice work! This is great
